File: program/program/find-the-maximum-sum-of-a-contiguous-subarray-(circular)/FindTheMaximumSumOfAContiguousSubarray(circular).ts
const maxSumSubarrayCircularTs: any = (nums: number[]) => {
  if (nums.length === 0) return 0;

  let sum = nums[0];

  let currMax: number = nums[0];
  let currMin: number = nums[0];
  let maxTotal: number = nums[0];
  let minTotal: number = nums[0];

  for (let i = 1; i < nums.length; i++) {
    let val: number = nums[i];

    currMax = Math.max(val, currMax + val);
    maxTotal = Math.max(currMax, maxTotal);

    currMin = Math.min(val, currMin + val);
    minTotal = Math.min(minTotal, currMin);

    sum += val;
  }

  if (sum == minTotal) return maxTotal;

  return Math.max(sum - minTotal, maxTotal);
};

var arr: number[] = [1, -2, 3, -2, 8, -5, 1, 7];
console.log(maxSumSubarrayCircularTs(arr));

File: program/program/find-the-maximum-sum-of-a-contiguous-subarray-(circular)/FindTheMaximumSumOfAContiguousSubarray(circular).cs
using System;
 
public class FindMaxContiguousSubarraySumCircular
{
    static int maxSubArraySumCircular(int[] a)
    {
        if (a.Length == 0) return 0;

        int sum = a[0];
        int currMax = a[0];
        int currMin = a[0];
        int maxTotal = a[0];
        int minTotal = a[0];

        for (int i = 1; i < a.Length; i++) {
            int val = a[i];

            currMax = val >= (currMax + val) ? val : (currMax + val);
            if (currMax > maxTotal)
                maxTotal = currMax;

            currMin = val < (currMin + val) ? val : (currMin + val);
            if (currMin < minTotal)
                minTotal = currMin;

            sum += val;
        }

        if (sum == minTotal)
            return maxTotal;

        return (sum - minTotal) > maxTotal ? (sum - minTotal) : maxTotal;
    }
     
    public static void Main ()
    {
        int[] a = {-2, -3, 4, -1, -2, 1, 5, -3};
        Console.Write("Max contiguous subarray sum (circular): " + maxSubArraySumCircular(a));
    }
}